### DISCUSSION TITLE
@narupley: Overwrite Logger Categories

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#5059] [Q&A]
Hi @narupley 

Thanks for your code snippet to overwrite logger categories:

https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Overwrite%20Logger%20Categories

The readme contains a hint: _Because of how we handle the filter/transformer, in order to get this to work in a transformer script you'll just have to add this line at the top of the first step in your transformer_

1. Can you explain why is that and maybe point me to the source code?
2. From line [9 to 19](https://github.com/nextgenhealthcare/connect-examples/blob/master/Code%20Templates/Overwrite%20Logger%20Categories/Overwrite%20Logger%20Categories.js#L9) you're defining an object containing all possible logger names as keys. Is it possible to get these names programmatically or are they hard-coded?

Many thanks!

**Comment by tom08zehn** [comment]
@narupley
Question #2: no more needed (not solved but different approach)
Question #1: still interested in

Many thanks.

**Reply by jonbartels** [reply]
> no more needed (not solved but different approach)

Can you say how you solved it? This helps the next person with questions.

> Because of how we handle the filter/transformer, in order to get this to work in a transformer script you'll just have to add this line at the top of the first step in your transformer

I'm not sure if I can explain this well but there are two possible explanations plus an experiment that might help show how it works:

1. Script locations other than transformers may load or operate differently. For example one executes implicitly but transformers have to be called manually
2. The function is essentially an override and script locations other than transformers call a function of the same name and this new definition overrides it. You could discover this by searching MC source in github for the function name.

A practical experiment might be to change the body of `overwriteCategories` in a test system then add a `throw`. That thrown exception should then log a stack trace showing the call stack to where it failed. For something like an explicit execution you'd see it thrown from some sort of "load" or "on-start" fall. For an override you'd see other code literally calling `overwriteCategories`.

**Reply by tom08zehn** [reply]
Q#1: I simply use the same renaming approach for all loggers and don't need a list of individual names in an object like `{ originalName1: "newName1", originalName2: "newName2" }`... I'm simply renaming all loggers to the naming convention `<channelName>[:<messageId> (if exists)]:<originalName>`.

code snippet:

```js
var originalName = logger.getName();
var newName = [channelName, originalName];
if(typeof connectorMessage !== "undefined") newName.splice(1, 0, connectorMessage.getMessageId()); // connectorMessage not exists in some scopes, e.g. deploy script
newName = newName.join(":");

var level = logger.getLevel();
logger = org.apache.log4j.Logger.getLogger(newName); // override happens here
logger.setLevel(level);
```

Q#2: And what is the reason why transformers are working different and are not calling the function from the code template lib even though it's available and callable in the transformer scope?

**Comment by tom08zehn** [comment]
I just wanted to share an update for Mirth v4.x using Log4j v2:

```js
const originalLoggerName = logger.getName();

var newName = [originalLoggerName];
if(typeof channelName !== "undefined") {
	if(typeof connectorMessage !== "undefined") newName.unshift(connectorMessage.getMessageId()); // connectorMessage not exists in some scopes, e.g. deploy script
	newName.unshift(channelName);
}
newName = newName.join(":");

// Get current log level
const level = logger.getLevel();
// Overwrite current logger with a new logger having a custom name
const logManager = org.apache.logging.log4j.LogManager;
logger = logManager.getLogger(newName);
// Set log level to what it was before
// Log4j **core** is not part of the public API, see https://stackoverflow.com/a/23434603
org.apache.logging.log4j.core.config.Configurator.setLevel(newName, level);
```
